- Description: dasd: Endless retry leads to busy / unresponsive system.
  Symptom:     Endless retry loop is blocking IO for a device.
  Problem:     Only ERP retries are using the retry counter (modify and check).
               All others (e.g. caused by the missing interrupt handler) did not.
               This sometimes leads to endless retry loops for this device.
  Solution:    Set and check the retry counter in start_IO for all requests.
  Problem-ID:  9795
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/block/dasd.c	2 Jul 2004 14:01:48 -0000	1.137.2.8
+++ linux-2.5/drivers/s390/block/dasd.c	2 Jul 2004 14:12:38 -0000	1.137.2.9
@@ -7,7 +7,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999-2001
  *
- * $Revision: 1.137.2.8 $
+ * $Revision: 1.137.2.9 $
  */
 
 #include <linux/config.h>
@@ -739,8 +739,16 @@
 	if (rc)
 		return rc;
 	device = (struct dasd_device *) cqr->device;
+	if (cqr->retries < 0) {
+		DEV_MESSAGE(KERN_DEBUG, device,
+			    "start_IO: request %p (%02x/%i) - no retry left.",
+			    cqr, cqr->status, cqr->retries);
+		cqr->status = DASD_CQR_FAILED;
+		return -EIO;
+	}
 	cqr->startclk = get_clock();
 	cqr->starttime = jiffies;
+	cqr->retries--;
 	rc = ccw_device_start(device->cdev, cqr->cpaddr, (long) cqr,
 			      cqr->lpm, 0);
 	switch (rc) {
@@ -1067,7 +1075,6 @@
 			break;
 		/*  Process requests with DASD_CQR_ERROR */
 		if (cqr->status == DASD_CQR_ERROR) {
-			cqr->retries--;
 			if (cqr->irb.scsw.fctl & SCSW_FCTL_HALT_FUNC) {
 				cqr->status = DASD_CQR_FAILED;
 				cqr->stopclk = get_clock();

